### PR TITLE
escapes regexy chars when using the flexMatcher

### LINKED ISSUE #1960: [psc-ide] Escape Regex Chars when flexmatching
psc-ide-server will crash if you pass a search string that contains "regexy" characters. Escape user input before turning it into the flexmatch regex.

### PR DESCRIPTION
**Pull request by kRITZCREEK**
This stops the server from crashing on queries like "<*"

fixes #1960

### CODE REVIEW

File: src/Language/PureScript/Ide/Matcher.hs [hunk 1]
```diff
@@ -86,15 +86,25 @@ flexRate pattern c@(Completion (_,ident,_)) = do
...
+        (-1,0) -> Nothing
+        (start,len) -> Just $ calcScore start (start + len)
+      where
+        escapedPattern :: [Text]
+        escapedPattern = map escape (T.unpack pattern)
+
+        -- escape wraps "regexy" characters in brackets to prevent the matcher
+        -- from crashing when trying to build the regex
+        escape :: Char -> Text
+        escape c = if c `elem` ("*.?$|\\" :: String)
```
Comment (michaelficarra) [inline src/Language/PureScript/Ide/Matcher.hs]: `^`, `(`, `)`, `{`, `}`, `[`, and `]` are all also meaningful regex characters.

File: src/Language/PureScript/Ide/Matcher.hs [hunk 2]
```diff
@@ -86,15 +86,25 @@ flexRate pattern c@(Completion (_,ident,_)) = do
...
+                   then "[" <> T.singleton c <> "]"
```
Comment (michaelficarra) [inline src/Language/PureScript/Ide/Matcher.hs]: You need to backslash escape these, since `^`, `]`, and `\` are meaningful inside a character class.

Comment (AppShipIt) [inline src/Language/PureScript/Ide/Matcher.hs]: Yep `[\^$.|?*+(){}`) ([reference](http://www.regular-expressions.info/refcharacters.html)).

Comment (kRITZCREEK) [inline src/Language/PureScript/Ide/Matcher.hs]: Does this also count for regex-tdfa? I don't think it implements the "real" perl regex spec.

Comment (AppShipIt) [inline src/Language/PureScript/Ide/Matcher.hs]: I'll look into it but IME using several implementations (mostly PCRE flavors) you can't escape too much.

**Comment by kRITZCREEK** [comment]
@michaelficarra did I get all of them? If yes i'll squash this.

**Comment by michaelficarra** [comment]
@kRITZCREEK Looks like it.

**Comment by kRITZCREEK** [comment]
Thanks! If we find more we can still add them. This is definitely an improvement over the status quo.

File: src/Language/PureScript/Ide/Matcher.hs [hunk 3]
```diff
@@ -86,15 +86,25 @@ flexRate pattern c@(Completion (_,ident,_)) = do
...
+        (-1,0) -> Nothing
+        (start,len) -> Just $ calcScore start (start + len)
+      where
+        escapedPattern :: [Text]
+        escapedPattern = map escape (T.unpack pattern)
+
+        -- escape prepends a backslash to "regexy" characters to prevent the
+        -- matcher from crashing when trying to build the regex
+        escape :: Char -> Text
+        escape c = if c `elem` ("[\\^$.|?*+(){}" :: String)
```
Comment (michaelficarra) [inline src/Language/PureScript/Ide/Matcher.hs]: What's with the explicit type signature here? Is that necessary?

Comment (garyb) [inline src/Language/PureScript/Ide/Matcher.hs]: `OverloadedStrings` is enabled in here I think, so probably.

Comment (kRITZCREEK) [inline src/Language/PureScript/Ide/Matcher.hs]: ^ Yep, that's why

**Comment by michaelficarra** [comment]
LGTM.

**Comment by paf31** [comment]
Thanks!

**Pull request merged**